src/Common/FileCache.cpp

```diff
@@ -0,0 +1,623 @@
+        queue.splice(queue.end(), queue, *cell.queue_iterator);
+    }
+}
+
+LRUFileCache::FileSegmentCell * LRUFileCache::getCell(
+    const Key & key, size_t offset, std::lock_guard<std::mutex> & /* cache_lock */)
+{
+    auto it = files.find(key);
+    if (it == files.end())
+        return nullptr;
+
+    auto & offsets = it->second;
+    auto cell_it = offsets.find(offset);
+    if (cell_it == offsets.end())
+        return nullptr;
+
+    return &cell_it->second;
+}
+
+FileSegments LRUFileCache::getImpl(
+    const Key & key, const FileSegment::Range & range, std::lock_guard<std::mutex> & cache_lock)
+{
+    /// Given range = [left, right] and non-overlapping ordered set of file segments,
+    /// find list [segment1, ..., segmentN] of segments which intersect with given range.
+
+    auto it = files.find(key);
+    if (it == files.end())
+        return {};
+
+    const auto & file_segments = it->second;
+    if (file_segments.empty())
+    {
+        removeFileKey(key);
+        return {};
+    }
+
+    FileSegments result;
+    auto segment_it = file_segments.lower_bound(range.left);
+    if (segment_it == file_segments.end())
+    {
+        /// N - last cached segment for given file key, segment{N}.offset < range.left:
+        ///   segment{N}                       segment{N}
+        /// [________                         [_______]
+        ///     [__________]         OR                  [________]
+        ///     ^                                        ^
+        ///     range.left                               range.left
+
+        const auto & cell = (--file_segments.end())->second;
+        if (cell.file_segment->range().right < range.left)
+            return {};
+
+        useCell(cell, result, cache_lock);
+    }
+    else /// segment_it <-- segmment{k}
+    {
+        if (segment_it != file_segments.begin())
+        {
+            const auto & prev_cell = std::prev(segment_it)->second;
+            const auto & prev_cell_range = prev_cell.file_segment->range();
+
+            if (range.left <= prev_cell_range.right)
+            {
+                ///   segment{k-1}  segment{k}
+                ///   [________]   [_____
+                ///       [___________
+                ///       ^
+                ///       range.left
+
+                useCell(prev_cell, result, cache_lock);
+            }
+        }
+
+        ///  segment{k} ...       segment{k-1}  segment{k}                      segment{k}
+        ///  [______              [______]     [____                        [________
+        ///  [_________     OR              [________      OR    [______]   ^
+        ///  ^                              ^                           ^   segment{k}.offset
+        ///  range.left                     range.left                  range.right
+
+        while (segment_it != file_segments.end())
+        {
+            const auto & cell = segment_it->second;
+            if (range.right < cell.file_segment->range().left)
```

Comment:
maybe add tiny helper lambda `isDisjoint`? Something like this, just for clarity.
https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreePartInfo.h#L61-L87

Reply:
No, because in half of the cases it is needed to check inclusion of only one boundary and not the second, but the function you advice to use checks both boundaries at once.